web/popminer/dispatch.go

```diff
@@ -0,0 +1,438 @@
+	}
+
+	compressedPubKey := privKey.PubKey().SerializeCompressed()
+	ethereumAddress := ethereum.PublicKeyToAddress(compressedPubKey).String()
+
+	return Object{
+		"ethereumAddress": ethereumAddress,
+		"network":         netNormalized,
+		"privateKey":      hex.EncodeToString(privKey.Serialize()),
+		"publicKey":       hex.EncodeToString(compressedPubKey),
+		"publicKeyHash":   btcAddress.AddressPubKeyHash().String(),
+	}.Value(), nil
+}
+
+func startPoPMiner(_ js.Value, args []js.Value) (js.Value, error) {
+	log.Tracef("startPoPMiner")
+	defer log.Tracef("startPoPMiner exit")
+
+	pmMtx.Lock()
+	defer pmMtx.Unlock()
+	if pm != nil {
+		return js.Null(), errors.New("pop miner already started")
+	}
+
+	pm = new(Miner)
+	pm.ctx, pm.cancel = context.WithCancel(context.Background())
+
+	cfg := popm.NewDefaultConfig()
+	cfg.BTCPrivateKey = args[0].Get("privateKey").String()
+	cfg.StaticFee = uint(args[0].Get("staticFee").Int())
+
+	cfg.LogLevel = args[0].Get("logLevel").String()
+	if cfg.LogLevel == "" {
+		cfg.LogLevel = "popm=ERROR"
+	}
+	loggo.ConfigureLoggers(cfg.LogLevel)
+
+	network := args[0].Get("network").String()
+	netOpts, ok := networks[network]
+	if !ok {
+		return js.Null(), fmt.Errorf("unknown network: %s", network)
+	}
+	cfg.BFGWSURL = netOpts.bfgURL
+	cfg.BTCChainName = netOpts.btcChainName
+
+	var err error
+	pm.miner, err = popm.NewMiner(cfg)
+	if err != nil {
+		return js.Null(), fmt.Errorf("create PoP miner: %w", err)
+	}
+
+	// launch in background
+	pm.wg.Add(1)
+	go func() {
+		defer pm.wg.Done()
+		if err := pm.miner.Run(pm.ctx); !errors.Is(err, context.Canceled) {
+			// TODO: dispatch event on failure
+			pmMtx.Lock()
+			defer pmMtx.Unlock()
+			pm.err = err // Theoretically this can logic race unless we unset om
+		}
+	}()
+
+	return js.Undefined(), nil
+}
+
+func stopPopMiner(_ js.Value, _ []js.Value) (js.Value, error) {
+	log.Tracef("stopPopMiner")
+	defer log.Tracef("stopPopMiner exit")
+
+	pmMtx.Lock()
+	if pm == nil {
+		pmMtx.Unlock()
+		return js.Null(), errors.New("pop miner not running")
+	}
+
+	oldPM := pm
+	pm = nil
+	pmMtx.Unlock()
+	oldPM.cancel()
+	oldPM.wg.Wait()
+
+	if oldPM.err != nil {
+		return js.Null(), oldPM.err
+	}
+
+	return js.Undefined(), nil
+}
+
+func ping(_ js.Value, _ []js.Value) (js.Value, error) {
+	log.Tracef("ping")
+	defer log.Tracef("ping exit")
+
+	activePM, err := activeMiner()
+	if err != nil {
+		return js.Null(), err
+	}
+	pr, err := activePM.miner.Ping(activePM.ctx, time.Now().Unix())
+	if err != nil {
+		return js.Null(), err
+	}
+
+	return Object{
+		"originTimestamp": pr.OriginTimestamp,
+		"timestamp":       pr.Timestamp,
+	}.Value(), nil
+}
+
+func l2Keystones(_ js.Value, args []js.Value) (js.Value, error) {
+	log.Tracef("l2Keystones")
+	defer log.Tracef("l2Keystones exit")
+
+	c := args[0].Get("numL2Keystones").Int()
+	if c < 0 || c > 10 {
+		c = 2
+	}
+	count := uint64(c)
+
+	activePM, err := activeMiner()
```

Comment:
I am not sure I love this construct. This is silly unless we actually have named miners that are passed in via args. Which is weird in its own special way.

This is very raceable.

Reply:
This behaviour is from the previous code, do you want it changed?